g$sam2gvcf <- function(inputfish = "", chrname = "", mem = 4, walltime = 72, recalibrate = TRUE,
		recalPlot = FALSE, Rversion = "3.1.2", GATKversion = "3.4.0", 
		genome = "gasAcu1pitx1new.fa", knownsites = "knownSnpsAllchrPitx1new.vcf", run = TRUE){
	# Takes .sam file all the way to gvcf file for a single fish.
	# Based on "sam2haplotypeCallerWithRecalibration.pbs" -- including arguments, for simplicity
	# RGPU (see below) is fixed as "Benlim-q15"
	# Recalibration is optional.
	# if recalPlot = TRUE, certain R libraries must be installed, see GATK help for AnalyzeCovariates.

	if(inputfish == "") stop("You need to provide inputfish")
	if(chrname != "") stop("Code not working for a single chromosome yet")
	if(genome != "gasAcu1pitx1new.fa") stop("Code currently works only with gasAcu1pitx1new.fa")
	if(knownsites != "knownSnpsAllchrPitx1new.vcf") stop("Code works only with knownSnpsAllchrPitx1new.vcf")

	# Attach date and time to name of pbs file to make unique
	hour <- gsub("[ :]", "-", Sys.time())
	pbsfile <- paste("sam2gvcf-", inputfish, chrname, "-", hour, ".pbs", sep = "")
	outfile <- file(pbsfile, "w")
	
	writeLines(			"#!/bin/bash", outfile)
	writeLines(			"#PBS -S /bin/bash", outfile)
	writeLines(paste(	"#PBS -l walltime=", walltime, ":00:00", sep=""), outfile)
	writeLines(paste(	"#PBS -l mem=", mem, "gb", sep = ""), outfile)

	writeLines(	"\n# pbs file to run sam2gvcf, based on 'sam2haplotypeCallerWithRecalibration.pbs'", outfile)
	writeLines(	  "# THIS FILE IS GENERATED BY R, DO NOT EDIT\n", outfile)
	
	writeLines("\necho \"Starting run at: \`date\`\"", outfile)

	writeLines('
	samfile="${inputfish}.sam"
	outfile="${inputfish}.out"
	RGID="BENLIM.${inputfish}"
	RGLB="${inputfish}.SB"
	RGSM="${inputfish}" 
	RGPL="ILLUMINA" 
	RGPU="Benlim-q15"
	fastafile="gasAcu1pitx1new.fa"
	faifile="gasAcu1pitx1new.fa.fai"
	dictfile="gasAcu1pitx1new.dict"
	intervals="${inputfish}.intervals"
	sortedbam="${inputfish}.sorted.bam"
	realignedbam="${inputfish}.realigned.bam"
	mkdupmetrics="${inputfish}.mkdup.metrics"
	mkdupbam="${inputfish}.mkdup.bam"
	mkdupbai="${inputfish}.mkdup.bai"
	reducedbam="${inputfish}.reduced.bam"
	vcffile="${inputfish}.vcf"
	recalbam="${inputfish}.recal.bam"
	recaltable="${inputfish}.recal.table"
	afterrecaltable="${inputfish}.after.recal.table"
	recalplots="${inputfish}.recal.pdf"
	
	module load samtools	
	', outfile)

	writeLines(paste('module load gatk', GATKversion, sep = "/"), outfile)
	writeLines(paste('module load R', Rversion, sep = "/"), outfile)

	writeLines('
	if [ ! -f "$dictfile" ]; then
	    java -Xmx2g -jar /global/software/picard-tools-1.89/CreateSequenceDictionary.jar R=$fastafile O=$dictfile
	fi
	
	if [ ! -f "$faifile" ]; then
	    samtools faidx $fastafile
	fi
	
	# 1. SortSam
	java -Xmx2g -jar /global/software/picard-tools-1.89/SortSam.jar I=$samfile O=$sortedbam \\
		SORT_ORDER=coordinate CREATE_INDEX=TRUE VALIDATION_STRINGENCY=LENIENT
	
	# 2. MarkDuplicates
	java -jar /global/software/picard-tools-1.89/MarkDuplicates.jar \\
		I=$sortedbam O=$mkdupbam M=$mkdupmetrics \\
		VALIDATION_STRINGENCY=LENIENT REMOVE_DUPLICATES=FALSE ASSUME_SORTED=TRUE
	
	# 3. AddOrReplaceReadGroups
	java -Xmx2g -jar /global/software/picard-tools-1.89/AddOrReplaceReadGroups.jar \\
		RGID=$RGID RGLB=$RGLB RGSM=$RGSM RGPL=$RGPL RGPU=$RGPU I=$mkdupbam O=$sortedbam \\
		SORT_ORDER=coordinate CREATE_INDEX=TRUE VALIDATION_STRINGENCY=LENIENT
	
	# 4. RealignerTargetCreator
	gatk.sh -Xmx4g -T RealignerTargetCreator -R $fastafile -I $sortedbam -o $intervals \\
		--allow_potentially_misencoded_quality_scores
	
	# 5. IndelRealigner:
	gatk.sh -Xmx4g -T IndelRealigner -R $fastafile -I $sortedbam -targetIntervals $intervals \\
		-o $realignedbam -LOD 0.4 --allow_potentially_misencoded_quality_scores
	', outfile)
	
	if(recalibrate){
		writeLines('
		# 6. BaseRecalibrator
		gatk.sh -Xmx4g -T BaseRecalibrator -R $fastafile -I $realignedbam \\
			-knownSites knownSnpsAllchrPitx1new.vcf -o $recaltable
		gatk.sh -Xmx4g -T PrintReads -R $fastafile -I $realignedbam -BQSR $recaltable -o $recalbam 
		', outfile)
		
		if(recalPlot){	
			writeLines('
			gatk.sh -Xmx4g -T BaseRecalibrator -R $fastafile -I $realignedbam \\
				-knownSites knownSnpsAllchrPitx1new.vcf -BQSR $recaltable -o $afterrecaltable
			gatk.sh -Xmx4g -T AnalyzeCovariates -R $fastafile -before $recaltable \\
				-after $afterrecaltable -plots $recalplots
			', outfile)
			}
		
		writeLines('
		# 7. HaplotypeCaller
		gatk.sh -Xmx4g -T HaplotypeCaller -R $fastafile -I $recalbam \\
		     --emitRefConfidence GVCF --variant_index_type LINEAR --variant_index_parameter 128000 \\
		      -o $vcffile --allow_potentially_misencoded_quality_scores
		', outfile)
	
		} else{
		writeLines('
		# 7. HaplotypeCaller
		gatk.sh -Xmx4g -T HaplotypeCaller -R $fastafile -I $realignedbam \\
		     --emitRefConfidence GVCF --variant_index_type LINEAR --variant_index_parameter 128000 \\
		      -o $vcffile --allow_potentially_misencoded_quality_scores
		', outfile)
		}

	writeLines('\nexit 0', outfile)

	writeLines('\necho \"Job finished with exit code $? at: \`date\`\"', outfile)
	
	close(outfile)
	
	# Use as "qsub -v inputfish=PRIL17 sam2haplotypeCallerWithRecalibration.pbs
	if(run){
		qsub <- paste("qsub -v inputfish=", inputfish, sep = "")
		system(paste(qsub, pbsfile))
		}
	}

g$qsubGenotypeGVCFsPbs <- function(gvcffiles, outvcfname, chromosome=NULL, GATK = "3.4.0", 
	mem = 4, walltime = 24, maxAltAlleles = 3, run = TRUE){
	# Generates qsub file "genotypeGVCFs.pbs" to carry out GATK genotypeGVCFs to call snps
	#	from multiple gvcf files inputted, one chromosome at a time
	# If chromosome is specified ("chrXXI" or "XXI") then only gvcf files having that chr are permitted.
	# If chromosome is not specified, it is extracted from gvcf file names.
	# Method assumes fasta file is named "chromosome.fa"
	# mem is in gb
	# walltime is in hours
	# Based on "makeGenotypeGVCFspbs.Rscript" but allows more flexibility in which files to analyze.
	
	cat("\ngvcffiles inputted:\n")
	for(i in 1:length(gvcffiles)){
		cat(gvcffiles[i],"\n")
		}

	# Get the chromosome number
	if(!is.null(chromosome)){
		chromosome <- gsub("chr","", chromosome)
		chromosome <- paste("chr", chromosome, sep = "")
		testfiles <- gvcffiles[grep(chromosome, gvcffiles)]
		if(length(testfiles) != length(gvcffiles)) stop("some filenames don't agree with chromosome name")
		}
	else{
		# Extract chromosome names
		z <- regexpr(text = gvcffiles, pattern = "[.]chr[A-z1]+[.]") # the "1" is for pitx1
		if( any(z < 1) ) stop("chromosome name missing from some filenames")
		chromosome <- substr(gvcffiles, z + 1, z + attr(z,"match.length") - 2)
		chromosome <- unique( chromosome )	
		# Check that there's only one chromosome included
		if( length(chromosome) != 1 ){
			stop("chr missing from some filenames, or more than one chromosome represented")
			}
		}

	# Attach date and time to name of file to make unique
	hour <- gsub("[ :]","-", Sys.time())
	pbsfilename <- paste("genotypeGVCFs", chromosome, hour, "pbs", sep = ".")

	pbsfile <- file(pbsfilename, "w")
	cat("\ninstructions being written to", pbsfilename, "\n")
	
	writeLines("#!/bin/bash", pbsfile)
	writeLines("#PBS -S /bin/bash", pbsfile)
#	writeLines("#PBS -l walltime=24:00:00", pbsfile)
	writeLines(paste("#PBS -l walltime=", walltime, ":00:00", sep = ""), pbsfile)
#	writeLines("#PBS -l mem=4gb", pbsfile)
	writeLines(paste("#PBS -l mem=", mem, "gb", sep = ""), pbsfile)
	writeLines("#PBS -l epilogue=epilogue.script", pbsfile)
	
	jobname <- gsub("[.]vcf", "", outvcfname)
	writeLines(paste("#PBS -N", jobname), pbsfile)
	writeLines("\n# THIS FILE IS GENERATED BY AN RSCRIPT, DO NOT EDIT\n", pbsfile)
	writeLines("# This code uses multiple vgcf files from HaplotypeCaller to calls snps", pbsfile)
	writeLines("# Run as \"qsub genotypeGVCFs.pbs\"\n", pbsfile)

	if( !all(grepl("[.]vcf$", gvcffiles)) ) stop("Provide only .vcf files as arguments")

	# We assume that the fasta file for the reference chromosome is named chr[A-z1].fa	
	fastafile <- paste(chromosome, ".fa", sep = "")
	if( !file.exists(fastafile) ) stop("'chr.fa' fasta file missing")
	
#	writeLines("module load gatk/3.4.0", pbsfile)
	writeLines(paste("module load gatk/", GATK, sep = ""), pbsfile)
#	writeLines(paste("# Using GATK v ", GATK, sep = ""), pbsfile)
	cat("\nUsing GATK v", GATK, "\n")
	
	writeLines("# the -jar GenomeAnalysisTK.jar argument is included in the java call in gatk.sh so superfluous here", pbsfile)

	# Print the main java command to the pbs file (\\ is to get single backslash)
#	writeLines(paste("gatk.sh -Xmx4g -R", fastafile, "-T GenotypeGVCFs \\"), pbsfile)
	writeLines(paste("gatk.sh -Xmx", mem, "g -R ", fastafile, " -T GenotypeGVCFs \\", sep = ""), pbsfile)
	vcfarguments <- gvcffiles
	for(i in 1:length(gvcffiles)){
		vcfarguments[i] <- paste("     --variant", vcfarguments[i], "\\")
		writeLines(vcfarguments[i], pbsfile)
		}
#	writeLines(paste("     --includeNonVariantSites --max_alternate_alleles 3 -o", outvcfname), pbsfile)
#	cat("Setting max_alternate_alleles to 3\n")
	writeLines(paste("     --includeNonVariantSites --max_alternate_alleles", maxAltAlleles, "-o", outvcfname), 
		pbsfile)
	cat("Uses --max_alternate_alleles", maxAltAlleles, "\n")
	
	close(pbsfile)
	if(run) system(paste("qsub", pbsfilename))
	}
