#!/usr/bin/Rscript

pbsfile <- file("genotypeGVCFs.pbs", "w")
writeLines("#!/bin/bash", pbsfile)
writeLines("#PBS -S /bin/bash", pbsfile)
writeLines("#PBS -l walltime=24:00:00", pbsfile)
writeLines("#PBS -l mem=4gb", pbsfile)
writeLines("#PBS -l epilogue=epilogue.script", pbsfile)

# Generates a qsub file "genotypeGVCFs.pbs" to carry out GATK's genotypeGVCFs to call snps from multiple gvcf files inputted.
# Can take a " *.chr??.vcf " wildcard as an argument and process all files accordingly, IGNORES CASE
# A specific chromosome name is extracted from file names.
# Run the qsub file after running this script.

# Run on hermes as:
#	/global/software/R-3.1.2/bin/Rscript makeGenotypeGVCFspbs.R pax*.chrxxi.vcf BenlimPaxton22fish.chrXXI.vcf

# arguments <- c(list.files(pattern=glob2rx("pax*.chrxxi.vcf"), ignore.case=TRUE), "BenlimPaxton22fish.chrXXI.vcf")
arguments <- commandArgs(TRUE) # only arguments will be stored
# print(arguments)

# in case a regular expression is first argument
fnames <- list.files(pattern = glob2rx(arguments[1]), ignore.case = TRUE)
arguments <- c(fnames, arguments[-1])
narg <- length(arguments)

# print(arguments) 
outvcfname <- arguments[narg]
gvcffiles <- arguments[-narg]
jobname <- gsub("[.]vcf", "", outvcfname)

writeLines(paste("#PBS -N", jobname), pbsfile)
writeLines("\n# THIS FILE IS GENERATED BY AN RSCRIPT, DO NOT EDIT\n", pbsfile)
writeLines("# This code combines vgcf files created by HaplotypeCaller and calls snps", pbsfile)
writeLines("# Run as \"qsub genotypeGVCFs.pbs\"\n", pbsfile)

if( !all(grepl("[.]vcf$", gvcffiles)) ) stop("Provide only .vcf files as arguments")

cat("\ngvcffiles inputted:\n")
for(i in 1:length(gvcffiles)){
	cat(gvcffiles[i],"\n")
	}

# Extract chromosome names
# z <- gsub("[A-z0-9_-]*[.](chr[A-z1]+)[.]vcf", "\\1", gvcffiles) # this works but without the chr check
z <- regexpr(text = gvcffiles, pattern = "[.]chr[A-z1]+[.]") # the "1" is for pitx1
if( any(z < 1) ){
	stop("'chr' missing from some filenames")
	}
chromosomes <- substr(gvcffiles, z + 1, z + attr(z,"match.length") - 2)
chromosomes <- unique( chromosomes )
# print(chromosomes)

# Check that there's only one chromosome included
if( length(chromosomes) != 1 ){
	stop("'chr' missing from some filenames or more than one chromosome included")
	}

# We assume that the fasta file for the reference chromosome is named chr[A-z1].fa	
fastafile <- paste(chromosomes, ".fa", sep = "")
cat("\nassumed name of reference chromosome:\n")
cat(fastafile,"\n")

writeLines("# Using GATK v 3.4.0", pbsfile)
writeLines("module load gatk/3.4.0", pbsfile)

writeLines("# the -jar GenomeAnalysisTK.jar argument is included in the java call in gatk.sh so superfluous here", pbsfile)

# Print the main java command to the pbs file (\\ is to get single backslash)
writeLines(paste("gatk.sh -Xmx4g -R", fastafile, "-T GenotypeGVCFs \\"), pbsfile)
vcfarguments <- gvcffiles

# Print multiple lines with all the gvcf file names
for(i in 1:length(gvcffiles)){
	vcfarguments[i] <- paste("     --variant", vcfarguments[i], "\\")
	writeLines(vcfarguments[i], pbsfile)
	}
# writeLines(paste("     --includeNonVariantSites --max_alternate_alleles 3 -o", outvcfname), pbsfile)
# cat("\nSetting max_alternate_alleles to 3\n")
writeLines(paste("     --includeNonVariantSites --max_alternate_alleles 3 -o", outvcfname), pbsfile)
cat("\nNot restricting max_alternate_alleles to 3\n")

# The following puts it all on one line instead
#for(i in 1:length(gvcffiles)){
#	vcfarguments[i] <- paste("--variant", vcfarguments[i])
#	}
#vcfarguments <- paste(vcfarguments, collapse = " ")
#vcfarguments <- paste(vcfarguments, "-o", outvcfname)
#writeLines(vcfarguments, pbsfile)

close(pbsfile)
